### PR TITLE
Ensure that symbolic variables incorporate fresh constraints before they're used

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87328
* #87263
* __->__ #87254
* #87252
* #87161

cc @jansel @lezcano @fdrocha

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -111,16 +111,24 @@ class PySymInt(object):
...
+        return self._expr
+
     def wrap(self, num):
         return PySymInt(sympy.Integer(num), self.shape_env, constant=num)
 
     def clone(self):
         return PySymInt(self.expr, self.shape_env, constant=self.constant)
 
+    def _update_expr(self):
+        self._expr = self.shape_env.simplify(self._expr)
```
Comment (bdhirsh) [inline torch/fx/experimental/symbolic_shapes.py:130]: do you think this will noticeably slow down compilation times 🤔. I hope `simplify()` is a fast no-op

Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:130]: We have a bunch of caches, so it'll never get run twice.

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -3801,25 +3801,6 @@ def forward(x):
         ]
         self.common(forward, args)
 
-    @unittest.skip("https://github.com/pytorch/torchdynamo/issues/1297")
```
Comment (Chillee) [inline test/inductor/test_torchinductor.py:3804]: Removing this since redundant test.

File: torch/_inductor/sizevars.py [hunk 3]
```diff
@@ -45,7 +45,7 @@ def __init__(self, shape_env=None):
         self.shape_env = shape_env
         self.var_to_val = self.shape_env.var_to_val
         self.guards = []
-        self.replacements: Dict[sympy.Symbol, Expr] = {}
+        self.replacements: Dict[sympy.Symbol, Expr] = self.shape_env.replacements
```
Comment (ezyang) [inline torch/_inductor/sizevars.py:48]: do you want an alias, or a copy?

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:130]: Not sure about that? We'd have to audit all expr call sites no?

**Comment by ezyang** [comment]
I'm not qualified to review the inductor code

File: torch/_inductor/sizevars.py (hunk 3 above)
Comment (Chillee) [reply torch/_inductor/sizevars.py:48]: Alias

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:130]: What do you mean? I'm saying that `self.shape_env.simplify` sits behind a LRU cache, so if we hit it twice it should mostly be a no-op.

File: torch/_inductor/sizevars.py (hunk 3 above)
Comment (ezyang) [reply torch/_inductor/sizevars.py:48]: might be worth a comment, as the Dict type annot is mutable

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**